This pull request adds the ECIES scheme and all required features

wusto opened the pull request.

Commit:
Add functions to import and export EC points

This patch adds the functionality to import and export EC public keys
using the EC points. Therefore two functions, one for importing one for
exporting, are added.

Commit:
Add OpenSSL functions for EC Point calculation

Commit:
Add ECIES scheme

Commit:
Fix integer comparison

blenk92 commented on src/kdf.cpp:
```diff
@@ -0,0 +1,135 @@
+X963KDF::X963KDF(const X963KDF &other)
+    : _impl(std::make_unique<X963KDF::Impl>(*other._impl)) {}
+
+X963KDF& X963KDF::operator=(X963KDF&& other) {
```
i guess { should be in the next line (at least seems to be the case everywhere else)

wusto replied:
done. See KDF-PR

blenk92 commented on src/key.cpp:
```diff
@@ -91,6 +91,42 @@ AsymmetricPublicKey AsymmetricPublicKey::readPublicKeyFromPEM(const std::string&
+    if (keySpec.get()->curve() == openssl::ellipticCurveNid::Ed448
+            || keySpec.get()->curve() == openssl::ellipticCurveNid::Ed25519)
+    {
+        throw MoCOCrWException("The given curve is not supported. Ed448 and Ed25519 is not supported");
```
```suggestion
        throw MoCOCrWException("The given curve is not supported. Ed448 and Ed25519 are not supported");
```

wusto replied:
done

blenk92 commented on src/key.cpp:
```diff
@@ -91,6 +91,42 @@ AsymmetricPublicKey AsymmetricPublicKey::readPublicKeyFromPEM(const std::string&
+    if (spec->curve() == openssl::ellipticCurveNid::Ed448
+            || spec->curve() == openssl::ellipticCurveNid::Ed25519)
+    {
+        throw MoCOCrWException("The given curve is not supported. Ed448 and Ed25519 is not supported");
```
```suggestion
        throw MoCOCrWException("The given curve is not supported. Ed448 and Ed25519 are not supported");
```

wusto replied:
done

blenk92 commented on src/key.cpp:
```diff
@@ -91,6 +91,42 @@ AsymmetricPublicKey AsymmetricPublicKey::readPublicKeyFromPEM(const std::string&
+            || spec->curve() == openssl::ellipticCurveNid::Ed25519)
+    {
+        throw MoCOCrWException("The given curve is not supported. Ed448 and Ed25519 is not supported");
+    }
```
maybe it makes sense to put the check for unsupported curves into a separate function (then we have to maintain the list of unsupported curves only in one place)

wusto replied:
Done

blenk92 commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     * 2. Encrypt the plaintext
+     * 3. Calculate the tag/mac (use getMAC() for getting it)
+     *
+     * After invoking this function, the ephemeral key and the authentication key can obtained using the getter
```
```suggestion
     * After invoking this function, the ephemeral key and the authentication key can be obtained using the getter
```

wusto replied:
Done

blenk92 commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     *
+     * The ephemeral key is generated during encryption. This public key (in combination with the private key of the
+     * receiver) is necessary for decrypting the ciphertext.
+     * @throws MoCOCrWException if finish() wasn't been invoked beforehand
```
```suggestion
     * @throws MoCOCrWException if finish() wasn't invoked beforehand
```

wusto replied:
Done

blenk92 commented on src/ecies.cpp:
```diff
@@ -0,0 +1,466 @@
+    checkForRequiredParameters();
+
+    ECIESEncryptionCtx *retVal;
+    if (_cert.internal()) {
```
i guess that should segfault if no _cert has been given. Maybe you want to use std::variant (not sure if we can already use it) or boost::variant to store pubkey / cert.

Or just check `_cert == nullptr`

wusto replied:
Cert is instantiated in the contstructor:
ECIESEncryptionCtxBuilder::ECIESEncryptionCtxBuilder(const AsymmetricPublicKey &bobsKey)
    : _pubKey(bobsKey), _cert(nullptr)
{}

ECIESEncryptionCtxBuilder::ECIESEncryptionCtxBuilder(const X509Certificate &bobsCert)
    : _pubKey(nullptr), _cert(bobsCert)
{}

Thus the X509 object always exists. (also check the header: "X509Certificate _cert;")
We should be able to check if the internal is either a nullptr or a "real" object.

garw replied:
yeah. See my comment above. This is a design flaw. The library was never intended to be used like this.

blenk92 replied:
ah i missed that it's not a pointer :-D

blenk92 commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+    std::function<std::unique_ptr<SymmetricCipherI> (const std::vector<uint8_t> &,
+                                                     const SymmetricCipherOperation&)> _symmetricCipherFactoryFunction;
+    std::vector<uint8_t> _kdfSalt;
+    std::vector<uint8_t> _macSalt;
```
i would like to have this class pimpled (same for the classes below).

wusto replied:
I don't prefer a pimpl pattern for builders.
@garw , @neverpanic : What is you opinion?

garw replied:
Well, I guess the major reason why we pimpl here is ABI stability. This probably applies for the builders as much as it does for the contexts, so I guess this is a valid point even though I agree that the maintainance effort is high.

neverpanic replied:
I agree with Martin, the PIMPL is for the ABI. It's annoying, but should really be done here.

wusto replied:
Done

blenk92 commented on src/ecies.cpp:
```diff
@@ -0,0 +1,466 @@
+        /* Copy first part (encryption key) */
+        std::vector<uint8_t>::const_iterator first = result.begin();
+        std::vector<uint8_t>::const_iterator last = result.begin() + _symCipherKeySize;
+        _key_encrypt = std::vector<uint8_t>(first, last);
```
```suggestion
        std::copy(result.begin(), result.begin() + _symCipherKeySize, _key_encrypt.begin());
```

or even nicer maybe (would be one copy less)

```
void createSymmetricKeys()
{
        std::vector<uint8_t> result = kdf->deriveKey(_sharedSecret, _macKeySize + _symCipherKeySize, _S1);
        std::copy(result.begin() + _symCipherKeySize, result.end(), _key_mac.begin()
        // add wipeout for bytes that will be removed here.
        result.resize(_symCipherKeySize);
        _key_encrypt = std::move(result);
```

garw replied:
max code example changes the semantics on how the key's are obtained from the KDF output and thus influence the overall algorithm. So, I guess this doesn't work here because it part of the specification from which part of the KDF output which key needs to come from.

wusto replied:
bought!

wusto replied:
Took the first approach.
Didn't find a solution to wipe the vector without a copy operation. So resizing doesn't work if you cannot wipe it before easily.

blenk92 commented on src/ecies.cpp:
```diff
@@ -0,0 +1,466 @@
+        /* Copy second part (mac key) */
+        first = result.begin() + _symCipherKeySize;
+        last = result.end();
+        _key_mac = std::vector<uint8_t>(first, last);
```
```suggestion
        std::copy(result.begin() + _symCipherKeySize, result.end(), _key_mac.begin());
```

wusto replied:
Done

blenk92 commented on src/ecies.cpp:
```diff
@@ -0,0 +1,466 @@
+        _key_mac = std::vector<uint8_t>(first, last);
+
+        /* Write zeros into result */
+        result.assign(result.size(), 0);
```
not sure if that may be optimized out

garw replied:
agreed. We probably need some kind of vectorShred(…) helper function that uses a secure wipe function from Openssl

wusto replied:
Also not sure. 
At least the gdb shows that it is overwritten by zeros...
@garw @neverpanic : Do you know if this is optimized out?

garw replied:
I guess this can depend on compiler version and compile options. Proofing that it is there in the current build is not really good enough. Hence, I would argue better safe than sorry. 
Compilers are able to detect memset(…) calls to dead memory and get rid of those…

neverpanic replied:
We have previously used `::explicit_bzero` for this (see src/symmetric_crypto.cpp), but OpenSSL may also have something here.

wusto replied:
Done.
I introduced a new utility function cleaning up vectors utilizing OPENSSL_cleanse.

blenk92 commented on src/ecies.cpp:
```diff
@@ -0,0 +1,466 @@
+        : ECIESImpl(kdf, macKeySize, macFactoryFunc, symCipherKeySize, symCipherFactoryFunc, S1, S2),
+          _ephemeralKey(nullptr)
+    {
+        const AsymmetricPublicKey pubKey{cert.getPublicKey()};
```
why not call the other Impl() Constructor. would make `initializer()` unnecessary

wusto replied:
good idea.

wusto replied:
Done

blenk92 commented on src/ecies.cpp:
```diff
@@ -0,0 +1,466 @@
+class ECIESEncryptionCtx::Impl : ECIESImpl
+{
+public:
+    explicit Impl(const AsymmetricPublicKey &pubKey,
```
explicit unneeded

wusto replied:
Done

blenk92 commented on src/ecies.cpp:
```diff
@@ -0,0 +1,466 @@
+    {
+        createEphemeralKeyAndSharedSecret(pubKey);
+        createSymmetricKeys();
+        /* Now we have generated the key for MAC -> generate it */
```
```suggestion
        /* Now we have generated the key for MAC -> create MAC instance*/
```

blenk92 commented on src/ecies.cpp:
```diff
@@ -0,0 +1,466 @@
+            const std::vector<uint8_t>&, const SymmetricCipherOperation&)> symCipherFactoryFunc,
+        const std::vector<uint8_t> &S1,
+        const std::vector<uint8_t> &S2)
+{
```
same as earlier, why not use the other `ECIESEncryptionCtx()` constructor

wusto replied:
done

blenk92 commented on src/ecies.cpp:
```diff
@@ -0,0 +1,466 @@
+    _impl->setMAC(tag);
+}
+
+ECIESCtxBuilder::ECIESCtxBuilder()
```
any particular reason to not use `= default` (if so, maybe worth a comment)

wusto replied:
Done

blenk92 commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+
+
+    /**
+     * @brief Sets the salt for the key derivation function (KDF). (Default: empty vector -> not used) (optional)
```
wouldn't be the sane default to autogenerate the salt?

garw replied:
Yeah, I am a bit unsure about this as well. Maybe let's have a chat how the ECIES variants handle this and if we really need this. If we don't right now, we might be better off the add this later when a use-case arises.

wusto replied:
@garw @neverpanic : imho this is a sane default, as the ephemeral key is the secret for ECIES and this one is generated every time you trigger the function.

garw replied:
The question is imho what the ECIES algorithm(s) specify.

wusto replied:
The ECIES algorithm specifies two additional parameters P1 and P2.
P1 = key deriv params
P2 = encoding parameters

In our case these variables are called 
kdfSalt -> P1
macSalt -> P2

blenk92 commented on tests/unit/test_ecies.cpp:
```diff
@@ -0,0 +1,186 @@
+    std::string sessionKey;
+};
+
+TEST(ECIESTest, deleteMe)
```
`deleteMe`?

wusto replied:
Done

blenk92 reviewed (dismissed):
Looks good in general (however, didn't look into the crypto yet). What I like to see in addition is some more tests for ECIES (for instace does it throw an exception if the some options are missing in the builder, maybe also tests with certificates and public keys,...)

neverpanic commented on src/mococrw/key.h:
```diff
@@ -89,6 +91,39 @@ class AsymmetricPublicKey
+     *
+     * This function creates a key object based on the compressed point in octet form
+     * and the elliptic curve group given by keySpec.
+     * The leading identifier (0x03 for compressed, 0x04 for uncompressed points. 0x07 for hybrid points) have to be in
```
```suggestion
     * The leading identifier (0x03 for compressed, 0x04 for uncompressed points. 0x07 for hybrid points) has to be in
```

wusto replied:
Done

neverpanic commented on src/mococrw/key.h:
```diff
@@ -89,6 +91,39 @@ class AsymmetricPublicKey
+     * The leading identifier (0x03 for compressed, 0x04 for uncompressed points. 0x07 for hybrid points) have to be in
+     * place.
+     *
+     * Remark: Ed-Curves (Ed448, Ed25519) are not supported. If you use one of these curves an MoCOCrWException is
```
I'd change this to:
```suggestion
     * Remark: Ed-Curves (Ed448, Ed25519) are currently not supported. If you use one of these curves an MoCOCrWException is
```

wusto replied:
Done

neverpanic commented on src/mococrw/key.h:
```diff
@@ -89,6 +91,39 @@ class AsymmetricPublicKey
+    /**
+     * @brief This returns the uncompressed points in octet representation of the given asymmetric public key.
+     *
+     * Depending of the form the correct identifier will be prepended to the actual point (0x03 for compressed,
```
Doesn't the sentence above indicate that this will always return the uncompressed version?

wusto replied:
Done.
We support compressed, uncompressed and hybrid points.

neverpanic commented on src/mococrw/key.h:
```diff
@@ -89,6 +91,39 @@ class AsymmetricPublicKey
+     * @throws MoCOCrWException if the given key is not a elliptic curve key or an Ed-key
+     * @param pubKey The public key to transform
+     * @param form The transformation form.
+     * @return The octet representation of the public key. The form of the point (
```
This sentence seems like you wanted to continue, but didn't?

wusto replied:
Seems like that :-D

neverpanic commented on src/openssl_wrap.cpp:
```diff
@@ -1289,5 +1289,36 @@ SSL_HMAC_CTX_Ptr _HMAC_CTX_new()
+    EC_KEY *key = OpensslCallPtr::callChecked(lib::OpenSSLLib::SSL_EVP_PKEY_get1_EC_KEY, evp_);
+    size_t length = OpensslCallIsPositive::callChecked(lib::OpenSSLLib::SSL_EC_KEY_key2buf, key,
+                                                       form, &pbuf, nullptr);
+    std::vector<uint8_t>result(pbuf, pbuf + length);
```
```suggestion
    std::vector<uint8_t> result(pbuf, pbuf + length);
```

wusto replied:
Done

neverpanic reviewed (changes requested):
Reviewed the first commit, more to come tomorrow.

garw commented on src/openssl_wrap.cpp:
```diff
@@ -1320,5 +1320,29 @@ std::vector<uint8_t> _EC_KEY_key2buf(const EVP_PKEY* evp, point_conversion_form_
+std::vector<uint8_t> _ECDH_compute_key(const EVP_PKEY *pub_key, const EVP_PKEY *ecdh,
+                                       void *(*KDF) (const void *in, size_t inlen, void *out, size_t *outlen))
+{
+    EVP_PKEY *pub_key_ = const_cast<EVP_PKEY*>(pub_key);
```
cost_cast always requires special care. It is sometimes needed because OpenSSL const correctness is not good but should always be checked in detail and a comment should confirm this/state why we believe that it is ok in this case.

wusto replied:
Done

garw commented on src/mococrw/util.h:
```diff
@@ -34,5 +36,17 @@ std::string toHex(const std::vector<uint8_t> &data);
 std::vector<uint8_t> cryptoRandomBytes(size_t length);
+
+template<typename T, typename U>
+std::unique_ptr<T> uniquePtrCast(std::unique_ptr<U> u)
```
should be an rvalue ref in parameter?
not sure how this actually works…

Also: name should reflect that this is a dynamic_cast. So maybe rather unique_ptr_dynamic_cast?

Moreover, this currently has the side-effect that the object will always be freed of the cast is not successful. Should be documented.

wusto replied:
Done

garw commented on src/mococrw/util.h:
```diff
@@ -34,5 +36,17 @@ std::string toHex(const std::vector<uint8_t> &data);
+        u.release();
+        return std::unique_ptr<T>(ptr);
+    }
+    throw MoCOCrWException("Can't cast unique pointer.");
```
Should also be documented. This is different compared to standard dynamic_cast, which doesn't throw

wusto replied:
Done

garw commented on src/mococrw/symmetric_crypto.h:
```diff
@@ -56,6 +56,12 @@ enum class SymmetricCipherPadding
     PKCS
 };
 
+enum class SymmetricCipherOperation
```
not sure if this is really needed. Let's see in a later commit. Just add a comment here so I don't forget

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+    /**
+     * @brief Finalize the encryption.
+     *
+     * This includes:
```
Clarify that these steps are done according to ECIES algorithm internally. Otherwise, this is probably confusing to a reader.

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     * This includes:
+     * 1. Creating the ephemeral key used for encryption (use getEphemeralKey() for getting it)
+     * 2. Encrypt the plaintext
+     * 3. Calculate the tag/mac (use getMAC() for getting it)
```
reference to getMAC is confusing. This is the public interface documentation. Do not state what you do internally on a detailed level!

wusto replied:
I'd like to point out to the user of the library, that the MAC can be obtained after invoking finish using the method getMAC().
Would you remove this here and put it to the getMAC() function?

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     * @throws MoCOCrWException if finish() wasn't invoked beforehand
+     * @return The tag/mac
+     */
+    std::vector<uint8_t> getMAC();
```
Why is this a separate function? Shouldn't it be placed at a position that is specified by this IEEE standard and returned by finalize()

neverpanic replied:
AFAIR the IEEE standard does not specify a serialization format, so we're implementing separate functions to get these values. This is similar to how, for example, Bouncy Castle implement this.

If we were to assume that ephemeral key, ciphertext and mac are just concatenated, getEphemeralKey() would also not make sense.

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+    std::vector<uint8_t> getMAC();
+
+private:
+    ECIESEncryptionCtx(const AsymmetricPublicKey& pubKey,
```
you could/should consider passing the builder object instead of two constructors with a huge list of parameters

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+
+private:
+
+    ECIESDecryptionCtx(const AsymmetricPrivateKey& privKey,
```
s.a.

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+{
+public:
+    /**
+     * @brief Sets the key derivation function used for ECIES (Default: mococrw::X963KDF with SHA512) (required)
```
How can it have a default and be required at the same time?

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     * @param kdf The actual object used in ECIES
+     * @return builder instance
+     */
+    ECIESCtxBuilder &setKDF(std::shared_ptr<mococrw::KeyDerivationFunction> kdf);
```
mixed handling: We expect unique_ptr in some and shated_ptr in other places. What's the general strategy? Is there a reason for the difference?

wusto replied:
The KeyDerivationFunction is stateless. Thus it is possible to have one instance used in multiple places.
It is also possible to change this to an unique_ptr. The interface will be more homogenous then.
Just drop me a comment what is favorable.

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     * This function is a dependency injection.
+     * (e.g. for HMAC-SHA512:
+     \code{.cpp}
+     auto macFunc = [](const std::vector<uint8_t> &key) {
+        std::unique_ptr<MessageAuthenticationCode>mac_ptr(new mococrw::HMAC(openssl::DigestTypes::SHA512, key, 64));
+        return mac_ptr;
+     };
```
Maybe better to read:
```suggestion
     auto macFunc = [](const std::vector<uint8_t> &key) -> std::unique_ptr<MessageAuthenticationCode>  {
        return new mococrw::HMAC(openssl::DigestTypes::SHA512, key, 64);
     };
```

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     };
+     \endcode
+     * )
+     * @param func A lambda expression which returns an object implementing the MessageAuthenticationCode interface
```
Doesn't need to be a lambda expression. Can be all kind of things that convert nicely into a std::function

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+            SymmetricCipherKeySize::S_256,
+            key).setIV(std::vector<uint8_t>(AESCipherBuilder::getDefaultIVLength(SymmetricCipherMode::CBC)));
+        builder.setPadding(SymmetricCipherPadding::PKCS);
+        if (operation == SymmetricCipherOperation::Decryption) {
```
Well, I guess here we go. This distinction shouldn't be needed as it is clear from the context if this is supposed to be used for encryption or decryption because you either put it into an Encryption or a DecryptionCtx. 

So, I would remove this here and also revert the change to that makes this enum public.

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+            key).setIV(std::vector<uint8_t>(AESCipherBuilder::getDefaultIVLength(SymmetricCipherMode::CBC)));
+        builder.setPadding(SymmetricCipherPadding::PKCS);
+        if (operation == SymmetricCipherOperation::Decryption) {
+            symCiphPtr = builder.buildDecryptor();
```
why do we have this intermediate variable? Just specify the return type of the lambda with the `->` notation as I illustrated above.

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     */
+    ECIESCtxBuilder &setSymmetricCipherFactoryFunction(std::function<std::unique_ptr<SymmetricCipherI>(
+                                                           const std::vector<uint8_t>&,
+                                                           const SymmetricCipherOperation&)> func);
```
As I wrote above. SymmetricCipherOperation doesn't make a lot of sense to me here. You know if you want to use the context for encryption or decryption?

neverpanic replied:
Same question as above, except in this case I'm wondering whether we need the flexibility this interface provides at all. Do we really want to support anything other than AES-256-CBC at the moment? Once we provide the flexibility, we can never remove it from the interface again, but if we don't provide it now we can always go back to provide it later and set the old implementation as default.

wusto replied:
The SymmetricCipherOperation is removed.
@neverpanic : I discussed it with Martin and we want to have it flexible at the moment.
Do  you want to speak to Martin directly about this?

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     * @brief Sets the salt for the message authentication code. (Default: empty vector -> not used). (optional)
+     * @param macSalt The salt value used for message authentication code.
+     * @return builder instance
+     */
```
same here?

wusto replied:
The IEEE-1363a standard proposes salts for MAC and KDF (named P1 and P2 in the standard).

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+
+};
+
+class ECIESEncryptionCtxBuilder : public ECIESCtxBuilder
```
This design breaks the nice way how one can ususally write down builders:
```
ECIESEncryptionBuilder(key)
  .setMacFactoryFunction(…)
  .setMacKeySize(…)
  .buildEncryptionCtx() // will now fail because the object here is not an ECIESEncryptionCtxBuilder.
```

If you want to provide a common implementation based on this base architecture, this is probably the time for the Curiously_recurring_template_pattern. We can discuss if you like.

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,466 @@
+class ECIESImpl
+{
+public:
+    ECIESImpl(const std::shared_ptr<KeyDerivationFunction> kdf,
```
Also here: Really consider passing the builder.

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,466 @@
+    virtual ~ECIESImpl()
+    {
+        /* Write zeros into _sharedSecret and both keys */
+        _key_encrypt.assign(_key_encrypt.size(), 0);
```
I would be concerned about compiler optimization here because it knows that these objects are about to be destroyed. Didn't we have something from openssl that securely deleted memory and circumvents compiler optimizations?

wusto replied:
Done. OPENSSL_clear_free is used instead.

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,466 @@
+        std::vector<uint8_t> result = kdf->deriveKey(_sharedSecret, _macKeySize + _symCipherKeySize, _S1);
+
+        /* Copy first part (encryption key) */
+        std::vector<uint8_t>::const_iterator first = result.begin();
```
really use auto in these cases. this is just super noisy code. If you want a const iterator, use cbegin()

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,466 @@
+        initializer(pubKey);
+    }
+
+    explicit Impl(const X509Certificate &cert,
```
There is delegating constructors in C++ since 11. However, if you just pass the builder, this might not be needed at all.

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,466 @@
+    void createEphemeralKeyAndSharedSecret(const AsymmetricPublicKey &pubKey)
+    {
+        // 1. Get the parameters from the public key and generate a new key based on it
+        AsymmetricKey key(pubKey.getKeySpec()->generate());
```
hmhm. Looking at the code, I would argue that the intention of the library author was that you use AsymmetricKeyPair here. this is then also a public key and can be used like one?

wusto replied:
Hmm. 
Somehow the public key interface is awkward to use. 
If I get the keySpec from a key I get a std::unique_ptr, but the generate Function takes a const ref on the keySpec...
--> AsymmetricKeypair key(AsymmetricKeypair::generate(*pubKey.getKeySpec().get()));

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,444 @@
+        _encryptor = _symCipherFactoryFunc(_key_encrypt, SymmetricCipherOperation::Encryption);
+    }
+
+    ~Impl() = default;
```
superfluous

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,444 @@
+
+};
+
+class ECIESDecryptionCtx::Impl : public ECIESImpl
```
public inheritance here. Default = private inheritance for the EncryptionCtx?

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,444 @@
+}
+
+ECIESEncryptionCtxBuilder::ECIESEncryptionCtxBuilder(const AsymmetricPublicKey &bobsKey)
+    : _pubKey(bobsKey), _cert(nullptr)
```
Please no. The certificate objects were never designed to be like an "optional". No code there is expecting a nullptr and I start wondering about the constructor from pointer being public. This seems in restrospective to be a bad idea. If you want something optional, make it explicitly optional. It's a design flaw in the library that this is possible in the first place.

wusto replied:
done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,444 @@
+{}
+
+ECIESEncryptionCtxBuilder::ECIESEncryptionCtxBuilder(const X509Certificate &bobsCert)
+    : _pubKey(nullptr), _cert(bobsCert)
```
same here. Key objects are also not designed to be an optional.

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,444 @@
+
+    ECIESEncryptionCtx *retVal;
+    if (_cert.internal()) {
+        retVal = new ECIESEncryptionCtx(_cert, _kdf, _macKeySize, _macFactoryFunction,
```
why not just return?

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,444 @@
+{
+    checkForRequiredParameters();
+
+    auto retVal = new ECIESDecryptionCtx(_privKey, _kdf, _macKeySize, _macFactoryFunction,
```
same here

wusto replied:
Done

garw commented on tests/unit/test_ecies.cpp:
```diff
@@ -0,0 +1,186 @@
+        };
+
+
+        encBuilder.setMacFactoryFunction(macFunc);
```
this doesn't look very builder-like

wusto replied:
Done

garw commented on tests/unit/test_ecies.cpp:
```diff
@@ -0,0 +1,186 @@
+        const SymmetricCipherPadding padding = SymmetricCipherPadding::PKCS;
+        const std::vector<uint8_t> zeroIV = std::vector<uint8_t>(AESCipherBuilder::getDefaultIVLength(mode));
+
+        kdf = std::shared_ptr<X963KDF>(new X963KDF(openssl::DigestTypes::SHA512));
```
use of std::make_shared is preferred. See Effective Modern C++ Item 21 on why

wusto replied:
Done

neverpanic commented on src/key.cpp:
```diff
@@ -26,6 +26,17 @@
 #include "mococrw/openssl_wrap.h"
 
+namespace {
+void isEccCurveSupported(mococrw::openssl::ellipticCurveNid nid)
```
Sort of a nitpick, feel free to ignore: Since this throws rather than returning a bool, I'd name this `ensureEccCurveSupported`.

wusto replied:
Done

neverpanic commented on src/mococrw/openssl_wrap.h:
```diff
@@ -1339,6 +1341,19 @@ enum class ellipticCurveNid
     Ed25519 = NID_ED25519
 };
 
+enum class ellipticCurvePointConversionForm
```
I haven't checked whether there's a precedent for this, but don't we usually capitalize types, i.e. should we make this `EllipticCurvePointConversionForm`?

neverpanic replied:
Friendly reminder?

wusto replied:
Won't change.

garw replied:
No will change! This is part of the public interface! Type names are capitilized.

neverpanic commented on src/mococrw/openssl_wrap.h:
```diff
@@ -1339,6 +1341,19 @@ enum class ellipticCurveNid
+enum class ellipticCurvePointConversionForm
+{
+    /* the point is encoded as z||x, where the octet z specifies
+     * which solution of the quadratic equation y is  */
```
```suggestion
     * which solution of the quadratic equation x is  */
```
Maybe I'm misunderstanding things, but shouldn't this be `x` because that's what you use in `z||x` above?

wusto replied:
Hmm. Good question. I copied it from the openssl header (include/openssl/ec.h). Perhaps the solution of the elliptic curve equation is meant (y^2 = x^3 + ax + b). But I am also not sure about this.

neverpanic replied:
Hmh, I guess if you just copied it, it should be fine for now.

neverpanic commented on src/mococrw/symmetric_crypto.h:
```diff
@@ -310,6 +310,13 @@ class AESCipherBuilder
+    /**
+     * @brief Get the default length of the IV in bytes given the cipher mode.
+     * @param mode The mode
+     * @return The lenght of the IV in bytes.
```
```suggestion
     * @return The length of the IV in bytes.
```

wusto replied:
Done

garw commented on src/openssl_wrap.cpp:
```diff
@@ -363,7 +363,7 @@ int _BIO_puts(BIO* bio, std::string buf)
 int _BIO_write(BIO* bio, const std::vector<uint8_t> &data)
 {
-    if (data.size() > std::numeric_limits<int>::max()) {
+    if (data.size() > static_cast<std::decay_t<decltype(data)>::size_type>(std::numeric_limits<int>::max())) {
```
Why this complicated? Readability counts, this is no template library. std::vector<uint8_t>::size_type is easily understandable to everyone if we want to be super future-proof. Otherwise, casting to size_t is probably already good enough.

neverpanic replied:
Sorry, I did this ;-)

wusto replied:
Done

garw commented on src/openssl_wrap.cpp:
```diff
@@ -376,7 +376,7 @@ int _BIO_write(BIO* bio, const std::vector<uint8_t> &data)
 int _BIO_read(BIO* bio, std::vector<uint8_t> &buf, std::size_t numBytes)
 {
-    if (numBytes > std::numeric_limits<int>::max()) {
+    if (numBytes > static_cast<decltype(numBytes)>(std::numeric_limits<int>::max())) {
```
Again, just size_t.

wusto replied:
Done

neverpanic commented on src/mococrw/symmetric_crypto.h:
```diff
@@ -56,6 +56,12 @@ enum class SymmetricCipherPadding
     PKCS
 };
 
+enum class SymmetricCipherOperation
```
Since this now part of the public interface, please add documentation.

wusto replied:
Done.
It's private again.

neverpanic commented on src/openssl_wrap.cpp:
```diff
@@ -1323,5 +1323,29 @@ std::vector<uint8_t> _EC_KEY_key2buf(const EVP_PKEY* evp, point_conversion_form_
 }
 
+
+std::vector<uint8_t> _ECDH_compute_key(const EVP_PKEY *pub_key, const EVP_PKEY *ecdh,
```
> Users of the OpenSSL library are expected to normally use the EVP method for working with Elliptic Curve Diffie Hellman as described above and on the EVP Key Agreement page. The EVP API is implemented by a lower level ECDH API. In some circumstances, expert users may need to use the low level API. *This is not recommended for most users.*

Am I correct in assuming that we're using the lower level API for the KDF support?

neverpanic replied:
Hm, you don't seem to be using the KDF support at all. In that case, I'd prefer if we used the higher-level API, as recommended by the OpenSSL wiki.

wusto replied:
Done.
I've just don't have a good name for it and think about continue to use ECDH_compute_key.

neverpanic commented on src/openssl_wrap.cpp:
```diff
@@ -1323,5 +1323,29 @@ std::vector<uint8_t> _EC_KEY_key2buf(const EVP_PKEY* evp, point_conversion_form_
+{
+    EVP_PKEY *pub_key_ = const_cast<EVP_PKEY*>(pub_key);
+    EVP_PKEY *ecdh_ = const_cast<EVP_PKEY*>(ecdh);
+    EC_KEY *ec_key = OpensslCallPtr::callChecked(lib::OpenSSLLib::SSL_EVP_PKEY_get1_EC_KEY, ecdh_);
```
Doesn't `EVP_PKEY_get1_EC_KEY` increase the reference count of the key and thus requires an associated free? `EC_KEY` does not seem to be a smart pointer that does this automatically (but maybe it should be)?

Did you run this under valgrind or with address sanitizer?

neverpanic replied:
The more I think about this the more I think our SSL_EVP_PKEY_get1_EC_KEY wrapper should return a smart pointer.

wusto replied:
Done

neverpanic commented on src/mococrw/openssl_wrap.h:
```diff
@@ -1374,5 +1374,7 @@ SSL_EC_KEY_Ptr _EC_KEY_oct2key(int nid, const std::vector<uint8_t> &buf);
 void _EVP_PKEY_set1_EC_KEY(EVP_PKEY *pkey, const EC_KEY *key);
 
 std::vector<uint8_t> _EC_KEY_key2buf(const EVP_PKEY* evp, point_conversion_form_t form);
+std::vector<uint8_t> _ECDH_compute_key(const EVP_PKEY *pub_key, const EVP_PKEY *ecdh,
```
Can we make the parameter names more expressive? I do get what the `pub_key` is, but I'm not sure why the private key is named `ecdh`.

I'd propose to use the nomenclature used by the OpenSSL code, which is `key` for the private key and `peerkey` for the public key of the other peer.

wusto replied:
Done

neverpanic commented on src/openssl_wrap.cpp:
```diff
@@ -1323,5 +1323,29 @@ std::vector<uint8_t> _EC_KEY_key2buf(const EVP_PKEY* evp, point_conversion_form_
+    EVP_PKEY *ecdh_ = const_cast<EVP_PKEY*>(ecdh);
+    EC_KEY *ec_key = OpensslCallPtr::callChecked(lib::OpenSSLLib::SSL_EVP_PKEY_get1_EC_KEY, ecdh_);
+    const EC_GROUP *ec_group = OpensslCallPtr::callChecked(lib::OpenSSLLib::SSL_EC_KEY_get0_group, ec_key);
+    EC_KEY *pub_ec_key = OpensslCallPtr::callChecked(lib::OpenSSLLib::SSL_EVP_PKEY_get1_EC_KEY, pub_key_);
```
Same issue here, the `EVP_PKEY_get1_EC_KEY` function increases the reference count and should come with an associated free.

wusto replied:
Done

garw reviewed (changes requested):
It took me quite a while to understand the abstraction level of what this is adding and I guess that documentation is needeed to enable other people to be able to understand this.
What I got from the code:

This implements an ECIES scheme in way that is similar to the IEEE standard with ECDH, key deriviation for symmetric cipher and a MAC method. The encryption and decryption contexts have nothing w.r.t serialization or deserialization of complete chunks of data.

So from my perspective, this needs to be explained for the {Encryption,Decryption}Ctx classes. I was quite confused for a while before I realized that the update() and finish() functions actually only handle the symmetric encryption. This is ok (in principle) but the expectation might be different when you look at an ECIES scheme. Which is the ECIES schemes that we have seen so far can be implemented from this? Only the IEEE variant? Also others? I mean, the idea is quite generic.
So an additional layer that also provides serialization and deserialization of the different components is needed. If this is not standardized, maybe as an example on what we would suggest. We can, however, and should resolve this patchsets first.

Some additional comments regarding the code:
* It feels very noisy to read in many places. In particular the huge parameter lists of the constructors are really bad for readability. In particular with names like `s1` or `s2`. In the caller they are still called kdfSalt and macSalt. this should be kept. Otherwise nobody can follow this implementation. You should consider passing the builders here because they already bundle everything. The impl classes can copy from there.
* omit the mococrw:: namespace prefix for implementation code that lives in the mococrw namespace anyways. This is just adding noise. 
* use auto instead of lengthy type names when the type can be inferred
* I saw in mutiple occasions the use of local variables that just serve the purpose of holding the return value until a return statement at the end of the function. Just return from where you want to return. It is way easier to read as it is clear that this will be returned.
* The use of make_unique and make_shared if preferred of raw new. See Effective Modern C++ Item 21 on why.
* In general, having a state in classes is adding complexity. If your class is stateful, an functionality overview in the documentation of this class should explain this.
* Comments (even documentation comments) that just repeat the obvious are discouraged.

neverpanic commented on src/openssl_wrap.cpp:
```diff
@@ -1323,5 +1323,29 @@ std::vector<uint8_t> _EC_KEY_key2buf(const EVP_PKEY* evp, point_conversion_form_
+
+    /* Calculate the public key and store it in result (Just the Px value (Py is omitted)) */
+    int actual_size = OpensslCallIsPositive::callChecked(lib::OpenSSLLib::SSL_ECDH_compute_key,
+                                                         static_cast<void*>(result.data()),result.size(),
```
```suggestion
                                                         static_cast<void*>(result.data()), result.size(),
```

wusto replied:
Done

neverpanic commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     * Encrypt a chunk of data.
+     *
+     * This function encrypts the data and stores it in the internal buffer.
+     * The data in the internal buffer is authenticated, when the finish method is invoked.
```
We're not authenticating data when encrypting, we're computing the auth tag so that the other party can later authenticate the ciphertext.

wusto replied:
Done

neverpanic commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     * Decrypt a chunk of data.
+     *
+     * This function stores the data in the internal buffer.
+     * The data in the internal buffer is decrypted, when the finish method is invoked.
```
This is an implementation detail and would change if we implement streaming decryption. Ideally, we would not have to change the documentation when implementing streaming decryption. Please write this as if the decryption happend in this method already.

wusto replied:
Done

neverpanic commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+    void update(const std::vector<uint8_t>& message);
+
+    /**
+     * @brief Decrypts all the content stored in the internal buffer via the update function.
```
Same here, this should be "finalizes decryption and verifies authenticity" or something similar.

wusto replied:
Done

neverpanic commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     * @param message chunk of data to decrypt.
+     * @throws MoCOCrWException if this function is invoked after finish was called.
+     */
+    void update(const std::vector<uint8_t>& message);
```
We should document (and enforce in the implementation) that users must have called setEphemeralKey before invoking this.

wusto replied:
Done

neverpanic commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     * @throws MoCOCrWException when the ephemeral public key has different domain parameters then the private key.
+     * @return The decrypted message
+     */
+    std::vector<uint8_t> finish();
```
We should document that users must have called `setMAC()` before invoking this.

wusto replied:
Done

neverpanic commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     * @param keySize The key size in bytes
+     * @return builder instance
+     */
+    ECIESCtxBuilder &setMacCipherKeySize(size_t keySize);
```
In the ECIESDecryptionCtx constructor you call this macKeySize, not macCipherKeySize. We should be consistent everywhere, and I prefer macKeySize. This also affects the documentation comment.

wusto replied:
Done

neverpanic commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     * @param func A lambda expression which returns an object implementing the MessageAuthenticationCode interface
+     * @return builder instance
+     */
+    ECIESCtxBuilder &setMacFactoryFunction(std::function<std::unique_ptr<MessageAuthenticationCode>(
```
Does the IEEE standard specify a default? If so, it would be nice if users didn't have to write this piece of boilerplate.

wusto replied:
The standard IEEC 1363a-2004 proposes KDF2 and MAC1.
The use case we want to support requires X9.63 and HMAC.
What do you think?

neverpanic commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,363 @@
+     *
+     * As the key for AES is generated during ECIES key generation, the AES object has to be instantiated after that.
+     * This function is a dependency injection.
+     * (e.g. for HMAC-SHA512:
```
Mentioning HMAC-SHA512 does not make sense here.

wusto replied:
I guess there is a misunderstanding. The following lines in the documentation would return an HMAC-SHA512 instance.

neverpanic commented on tests/unit/test_ecies.cpp:
```diff
@@ -0,0 +1,186 @@
+        decBuilder.setSymmetricCipherKeySize(256 / 8);
+
+        encCtx = encBuilder.buildEncryptionCtx();
+        decCtx = decBuilder.buildDecryptionCtx();
```
Shouldn't most of this stuff be the default? We don't want our users to have to specify 5 parameters for *every* use of an ECIES context.

wusto replied:
Done

neverpanic commented on tests/unit/test_ecies.cpp:
```diff
@@ -0,0 +1,186 @@
+
+    decCtx->setMAC(mac);
+    decCtx->setEphemeralKey(ephKey);
+    decCtx->update(ciphertext);
```
Not the order this would be done in during streaming decryption. Please use setEphemeralKey, followed by update, then setMAC, then finish.

wusto replied:
Done

neverpanic commented on tests/unit/test_ecies.cpp:
```diff
@@ -0,0 +1,186 @@
+    std::string sessionKey;
+};
+
+TEST(ECIESTest, testWithBackendData)
```
Not quite sure this is a suitable test name for an open source project.

wusto replied:
Done

neverpanic commented on tests/unit/test_ecies.cpp:
```diff
@@ -0,0 +1,186 @@
+    const std::vector<uint8_t> zeroIV = std::vector<uint8_t>(AESCipherBuilder::getDefaultIVLength(mode));
+
+    /* Create KDF context */
+    auto kdf = std::make_shared<X963KDF>(X963KDF(openssl::DigestTypes::SHA512));
```
Isn't this the same as

```suggestion
    auto kdf = std::make_shared<X963KDF>(openssl::DigestTypes::SHA512);
```

wusto replied:
Done

neverpanic commented on tests/unit/test_ecies.cpp:
```diff
@@ -0,0 +1,186 @@
+    /* Create the decryption context */
+    decBuilder.setKDF(kdf)
+              .setMacFactoryFunction([](const std::vector<uint8_t> &key) {
+                    return std::make_unique<mococrw::HMAC>(mococrw::HMAC(openssl::DigestTypes::SHA512, key));
```
Same here:
```suggestion
                    return std::make_unique<mococrw::HMAC>(openssl::DigestTypes::SHA512, key);
```

wusto replied:
Done

neverpanic commented on tests/unit/test_ecies.cpp:
```diff
@@ -0,0 +1,186 @@
+    auto decCtx = decBuilder.buildDecryptionCtx();
+
+    /* Set the received mac value */
+    decCtx->setMAC(utility::fromHex(data.macTag));
```
Same thing here, please use setEphemeralKey, update, setMAC, finish in this order.

wusto replied:
Done

neverpanic commented on src/ecies.cpp:
```diff
@@ -0,0 +1,444 @@
+protected:
+    void createSymmetricKeys()
+    {
+        std::vector<uint8_t> result = kdf->deriveKey(_sharedSecret, _macKeySize + _symCipherKeySize, _S1);
```
_sharedSecret is only ever used in this function, and this function is always called after setting _sharedSecret.

Maybe _sharedSecret should not exist at all and be a parameter of this function? I don't see a good reason to keep it around after setting _key_encrypt and _key_mac.

The same thing actually applies for _key_encrypt and _key_mac, they are only used after calling this function to create the MAC and the symmetric encryption context.

Can we turn this into something similar to:
```c++
sharedSecret = ECDH(...);
auto keys = createSymmetricKeys(std::move(sharedSecret));
_mac = createMAC(std::move(keys.macKey));
_encryptor = createEncryptor(std::move(keys.encryptionKey));
```

This would also allow us to not care about secure deletion of the shared secret, mac key and encryption key from this class, because we could just delegate this to the MAC and symmetric crypto implementations if we never keep a copy of the key around.

wusto replied:
Done

Commit:
Fix for HMAC: change to call by reference

Commit:
Add utility for vector clean up

When vectors are destructed the content is not overwritten with zeros.
As some of the vectors we are using contain sensitive data, we need a
wrapper for vectors, which is overwritting the content during
destruction.

Commit:
symmetric-encryption: Add getter functions of IV length and key size

Two functions are changed/added:
* Make IV length getter public
* Added new getter for key size

neverpanic commented on src/ecies.cpp:
```diff
@@ -0,0 +1,497 @@
+        });
+
+        setMacKeySize(512 / 8);
+
```
```suggestion
```

Nitpick.

wusto replied:
Done

neverpanic commented on src/ecies.cpp:
```diff
@@ -0,0 +1,497 @@
+        const std::vector<uint8_t> zeroIV = std::vector<uint8_t>(AESCipherBuilder::getDefaultIVLength(mode));
+
+        if (isEncryption) {
+            setSymmetricCipherFactoryFunction([zeroIV](const std::vector<uint8_t> &key) -> std::unique_ptr<AESCipher> {
```
Not quite sure how you can use `mode`, `keySize` and `padding` here without having them captured in the lambda, but I guess if this compiles, it's fine.

wusto replied:
If I add it, I get the following warning:
ecies.cpp:73:48: warning: lambda capture 'mode' is not required to be captured for this use
-Wunsued-lambda-capture

neverpanic commented on src/ecies.cpp:
```diff
@@ -0,0 +1,497 @@
+        if (!_kdf) {
+            setDefaultKdf();
+        }
+        if ((_macFactoryFunction && !_macKeySize) || (!_macFactoryFunction && _macKeySize)) {
```
Not sure whether that would be an improvement of readability, but you could write this as an explicit XOR, i.e.
```cpp
if (_macFactoryFunction ^ _macKeySize)
```

This may require explicit `static_cast<bool>`s, though, in which case the readability improvement would immediately be out of the window again.

wusto replied:
Also not sure about this. At least in that way you can see what is actually checked.
I just wait for other comments :)

neverpanic commented on src/ecies.cpp:
```diff
@@ -0,0 +1,497 @@
+
+    }
+
+    const std::shared_ptr<KeyDerivationFunction> kdf;
```
```suggestion
    const std::shared_ptr<KeyDerivationFunction> _kdf;
```

wusto replied:
Done

neverpanic commented on src/ecies.cpp:
```diff
@@ -0,0 +1,497 @@
+        std::copy(result.begin() + _symmetricKeySize, result.end(), _key_mac.begin());
+
+        /* Write zeros into result */
+        utility::vectorCleanse(result);
```
This isn't required, since `result_deleter` will do this automatically, no?

wusto replied:
You are right. Forgot to remove this.

neverpanic commented on src/ecies.cpp:
```diff
@@ -0,0 +1,497 @@
+    bool _isFinished = false;
+};
+
+class ECIESEncryptionCtx::Impl : ECIESImpl
```
Shouldn't this also be public inheritance like you did it for `ECIESDecryptionCtx`?

```suggestion
class ECIESEncryptionCtx::Impl : public ECIESImpl
```

wusto replied:
Done

neverpanic commented on src/ecies.cpp:
```diff
@@ -0,0 +1,497 @@
+
+        std::vector<uint8_t> result;
+
+        /* Finish the encryption */
```
```suggestion
        /* Finish the decryption */
```

wusto replied:
Done

neverpanic commented on src/ecies.cpp:
```diff
@@ -0,0 +1,497 @@
+
+ECIESCtxBuilder &ECIESCtxBuilder::setKDFSalt(std::vector<uint8_t> kdfSalt)
+{
+    _impl->setKDFSalt(kdfSalt);
```
```suggestion
    _impl->setKDFSalt(std::move(kdfSalt));
```

wusto replied:
Done

neverpanic commented on src/ecies.cpp:
```diff
@@ -0,0 +1,497 @@
+
+ECIESCtxBuilder &ECIESCtxBuilder::setMACSalt(std::vector<uint8_t> macSalt)
+{
+    _impl->setMACSalt(macSalt);
```
```suggestion
    _impl->setMACSalt(std::move(macSalt));
```

wusto replied:
Done

neverpanic commented on src/ecies.cpp:
```diff
@@ -0,0 +1,497 @@
+
+std::unique_ptr<ECIESEncryptionCtx> ECIESCtxBuilder::buildEncryptionCtx(const AsymmetricPublicKey &bobsKey)
+{
+    _impl->_pubKey = bobsKey;
```
Probably not a good idea to store a reference to the key in the builder, since we don't know what the lifetime of the passed reference will be.

Since `AssymmetricPublicKey` is just a very lightweight wrapper around a shared pointer, I think it makes sense to just accept a `AsymmetricPublicKey` object here, and ideally even `std::move()` it into `_impl->_pubKey`.

wusto replied:
Done

neverpanic commented on src/ecies.cpp:
```diff
@@ -0,0 +1,497 @@
+        setSymmetricCipherKeySize(256 / 8);
+    }
+
+    void checkForRequiredParameters(bool isEncryption)
```
Boolean parameters are always somewhat weird, because you always have to read the documentation to figure out what they mean. I would have preferred an enum here, e.g. `checkForRequiredParameters(Mode::Encryption)`, which would be much easier to read.

wusto replied:
Done

neverpanic commented on src/ecies.cpp:
```diff
@@ -0,0 +1,497 @@
+
+std::unique_ptr<ECIESDecryptionCtx> ECIESCtxBuilder::buildDecryptionCtx(const AsymmetricPrivateKey &bobsKey)
+{
+    _impl->_privKey = bobsKey;
```
The same would apply here.

wusto replied:
Done

neverpanic commented on src/ecies.cpp:
```diff
@@ -0,0 +1,497 @@
+std::unique_ptr<ECIESEncryptionCtx> ECIESCtxBuilder::buildEncryptionCtx(const X509Certificate &bobsCert)
+{
+    return buildEncryptionCtx(bobsCert.getPublicKey());
+}
+
+std::unique_ptr<ECIESDecryptionCtx> ECIESCtxBuilder::buildDecryptionCtx(const AsymmetricPrivateKey &bobsKey)
+{
+    _impl->_privKey = bobsKey;
+    _impl->checkForRequiredParameters(false);
+
+    /* We cannot use std::make_unique because:
+     * - The constructor of ECIESEncryptionCtx is private
+     * - The ECIESEncryptionCtxBuilder is a friend of ECIESEncryptionCtx
+     * - If we call std::make_unique (which is a function call), we invoke the constructor from outside of the
+     *   ECIESEncryptionCtxBuilder. Thus the call is not allowed as it is private (friend is not transitive)
```
`s/EncryptionCtx/DecryptionCtx/g`?

wusto replied:
Done

neverpanic commented on src/key.cpp:
```diff
@@ -91,6 +102,30 @@ AsymmetricPublicKey AsymmetricPublicKey::readPublicKeyFromPEM(const std::string&
     return AsymmetricPublicKey{std::move(key)};
 }
 
+AsymmetricPublicKey AsymmetricPublicKey::fromECPoint(const std::shared_ptr<ECCSpec>keySpec,
```
```suggestion
AsymmetricPublicKey AsymmetricPublicKey::fromECPoint(const std::shared_ptr<ECCSpec> keySpec,
```

wusto replied:
Done

neverpanic commented on src/mococrw/key.h:
```diff
@@ -89,6 +91,39 @@ class AsymmetricPublicKey
+     * @param form The transformation form.
+     * @return The octet representation of the public key. The form of the point is defined by the argument form.
+     */
+    static std::vector<uint8_t> toECPoint(AsymmetricPublicKey &pubKey, openssl::ellipticCurvePointConversionForm form);
```
Why is this static? Wouldn't it make more sense to make this a member function and operate on the instance?

wusto replied:
Jep. For this function definitive. It's most probably a copy pasta failure.

neverpanic commented on src/key.cpp:
```diff
@@ -91,6 +102,30 @@ AsymmetricPublicKey AsymmetricPublicKey::readPublicKeyFromPEM(const std::string&
+std::vector<uint8_t> AsymmetricPublicKey::toECPoint(AsymmetricPublicKey &pubKey, ellipticCurvePointConversionForm form)
+{
+    std::shared_ptr<AsymmetricKey::Spec> keySpec = pubKey.getKeySpec();
+    auto spec = std::dynamic_pointer_cast<ECCSpec>(keySpec);
```
Since this is a downcast, you must check its result. Invoking this with an RSA public key will currently crash with a nullptr dereference when calling `spec->curve()`.

Please add
```cpp
if (!spec) {
  // error handling
}
```

wusto replied:
Done

neverpanic commented on src/key.cpp:
```diff
@@ -91,6 +102,30 @@ AsymmetricPublicKey AsymmetricPublicKey::readPublicKeyFromPEM(const std::string&
+    auto spec = std::dynamic_pointer_cast<ECCSpec>(keySpec);
+    ensureEccCurveSupported(spec->curve());
+
+    std::vector<uint8_t> result(_EC_KEY_key2buf(pubKey.internal(), static_cast<point_conversion_form_t>(form)));
+
+    return result;
```
```suggestion
    return _EC_KEY_key2buf(pubKey.internal(), static_cast<point_conversion_form_t>(form)));
```

wusto replied:
Done

neverpanic commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,402 @@
+ * If DH or DHC is used depends on the public key.
+ *
+ * Remark:
+ * - ECSVP-DHC (Diffie Helman with cofactor exponentiation/multiplication): Currently we don't support curves that have
```
```suggestion
 * - ECSVP-DHC (Diffie Hellman with cofactor exponentiation/multiplication): Currently we don't support curves that have
```

wusto replied:
Done

neverpanic commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,402 @@
+ * Remark:
+ * - ECSVP-DHC (Diffie Helman with cofactor exponentiation/multiplication): Currently we don't support curves that have
+ *   a cofactor different to h = 1. Therefor DHC is "supported" implicitly. As soon as curves having a cofactor h != 1
+ *   are needed the implementation have to be adapted.
```
```suggestion
 *   are needed the implementation has to be adapted.
```

wusto replied:
Done

neverpanic commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,402 @@
+ *   are needed the implementation have to be adapted.
+ *
+ * Elliptic curve integrated encryption scheme provides you the functionality to encrypt a plaintext with an one time
+ * shared secret (based on a ephemeral key) using a symmetric encryption scheme. All you have to know is:
```
```suggestion
 * shared secret (based on an ephemeral key) using a symmetric encryption scheme. All you have to know is:
```

wusto replied:
Done

neverpanic commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,402 @@
+ * 3. calculate and verify the MAC (using the optional salt for MAC)
+ * 4. decrypt the ciphertext
+ *
+ * Send P_eph, ciphertext and MAC to the recipient
```
This sentence does not make sense for the decryption context.

wusto replied:
That's right. :)

neverpanic commented on src/openssl_wrap.cpp:
```diff
@@ -1292,5 +1292,76 @@ SSL_HMAC_CTX_Ptr _HMAC_CTX_new()
+
+void _EVP_PKEY_set1_EC_KEY(EVP_PKEY *pkey, const EC_KEY *key)
+{
+    EC_KEY *key_ = const_cast<EC_KEY*>(key);
```
Please add a comment here explaining that the const correctness of the OpenSSL interface is wrong, and we're fixing it at this layer.

wusto replied:
Done

neverpanic commented on src/openssl_wrap.cpp:
```diff
@@ -1292,5 +1292,76 @@ SSL_HMAC_CTX_Ptr _HMAC_CTX_new()
+
+std::vector<uint8_t> _EC_KEY_key2buf(const EVP_PKEY* evp, point_conversion_form_t form)
+{
+    EVP_PKEY *evp_ = const_cast<EVP_PKEY*>(evp);
```
The same applies here.

wusto replied:
Done

neverpanic commented on tests/unit/test_ecies.cpp:
```diff
@@ -0,0 +1,382 @@
+{
+    /* Testing for missing parameters. This is the same for Encryption and decryption */
+    ECIESCtxBuilder decBuilder;
+    decBuilder.buildDecryptionCtx(randomKey);
```
If this is an incomplete builder, shouldn't this throw?

wusto replied:
Done

neverpanic commented on tests/unit/test_ecies.cpp:
```diff
@@ -0,0 +1,382 @@
+
+    auto kdf = std::make_shared<X963KDF>(openssl::DigestTypes::SHA512);
+
+    auto cipherDecFunc = [zeroIV](const std::vector<uint8_t> &key) -> std::unique_ptr<AESCipher> {
```
Wouldn't the defaults for all of these parameters be sufficient in this test?

wusto replied:
Done

wusto reviewed (commented):
Resolved the comments.
Will do a force push now

neverpanic commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+namespace  {
+enum class Mode {
+    Decryption = 0,
+    Encrytption
```
```suggestion
    Encryption
```

wusto replied:
Done

neverpanic commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,400 @@
+     * @param bobsKey The public key of the receiver
+     * @return An ECIES encryption context, which can be used for encryption.
+     */
+    std::unique_ptr<ECIESEncryptionCtx> buildEncryptionCtx(const AsymmetricPublicKey bobsKey);
```
I guess since we now grab a copy of this, this no longer needs to be const?

wusto replied:
Done

neverpanic commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,400 @@
+     * @param bobsKey The private key of the receiver
+     * @return An ECIES decryption context, which can be used for decryption.
+     */
+    std::unique_ptr<ECIESDecryptionCtx> buildDecryptionCtx(const AsymmetricPrivateKey bobsKey);
```
Same here.

wusto replied:
Done

neverpanic commented on src/mococrw/key.h:
```diff
@@ -89,6 +91,39 @@ class AsymmetricPublicKey
+     * Remark: Ed-Curves (Ed448, Ed25519) are currently not supported. If you use one of these curves a
+     * MoCOCrWException is thrown.
+     * @throws MoCOCrWException if the given key is not a elliptic curve key or an Ed-key
+     * @param pubKey The public key to transform
```
```suggestion
```

wusto replied:
Done

neverpanic reviewed (changes requested):
Here are some more nitpicks.

neverpanic reviewed (approved):
lg2m, @garw should probably take another look.

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+
+    void setKDF(std::shared_ptr<KeyDerivationFunction> kdf)
+    {
+        _kdf = kdf;
```
can be moved

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+                return AESCipherBuilder(mode, keySize, key).setIV(zeroIV).setPadding(padding).buildEncryptor();
+            });
+        } else {
+            setSymmetricCipherFactoryFunction([zeroIV](const std::vector<uint8_t> &key) -> std::unique_ptr<AESCipher> {
```
SymmetricCipherI instead of AESCipher here?

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+            setSymmetricCipherFactoryFunction([zeroIV](const std::vector<uint8_t> &key) -> std::unique_ptr<AESCipher> {
+                return AESCipherBuilder(mode, keySize, key).setIV(zeroIV).setPadding(padding).buildEncryptor();
+            });
+        } else {
```
ok here but in general 
```
if ()
else if ()
else if ()
else { error } 
```
is preferred for enums to avoid hidden bugs when enum elements are added.

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+            });
+        }
+
+        setSymmetricCipherKeySize(256 / 8);
```
should depend on the keySize. We should add a function to get the actual key size from the enum entry if this doesn't already exist.

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+        : _kdf(ctxBuilder._impl->_kdf),
+          _macKeySize(ctxBuilder._impl->_macKeySize),
+          _macFactoryFunc(ctxBuilder._impl->_macFactoryFunction),
+          _mac(nullptr),
```
not needed smart ptr intialize automatically to nullptr

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+          _symCipherFactoryFunc(ctxBuilder._impl->_symmetricCipherFactoryFunction),
+          _kdfSalt(ctxBuilder._impl->_kdfSalt),
+          _macSalt(ctxBuilder._impl->_macSalt),
+          _symcryptor(nullptr)    {}
```
s.a. and other occurences

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+        });
+
+        std::vector<uint8_t> result;
+        utility::Finally result_deleter([&result]() {
```
Took me a bit to verify that this is actually ok but local parameters are destroyed in the reverse order of their construction and parameters are constructed/destructed in the context of the caller. Should thus be fine.

Wouldn't hurt to state this here to indicate that this was checked.

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+    const std::vector<uint8_t> _macSalt;
+
+    std::unique_ptr<SymmetricCipherI> _symcryptor;
+    std::vector<uint8_t> _mac_value;
```
_ notation vs CamelCase. This library usually uses CamelCase.

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+    const std::vector<uint8_t> _kdfSalt;
+    const std::vector<uint8_t> _macSalt;
+
+    std::unique_ptr<SymmetricCipherI> _symcryptor;
```
we have code completion. Just call it symmetricCipher. I was a bit confused when seing symcryptor in the parameter list. Not really a self-explainatory name.

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+
+        std::vector<uint8_t> result;
+
+        /* Finish the encryption */
```
Comments that just repeat the code are discouraged. 
A statement about the actual algorithm would be appreciated: Something like: IEEE xyz specifies that the MAC will be computed over the ciphertext appended with the macSalt.

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+
+    AsymmetricPublicKey getEphemeralKey()
+    {
+        /* Return the ephemeral key */
```
s.a. and other occurences

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+         _privKey(ctxBuilder._impl->_privKey)
+    {}
+
+    void update(const std::vector<uint8_t>& message)
```
It seems a bit that nothing is possible here before the ephemeral key is set. So why isn't this part of the builder or handed to constructor? Seems like a superflous additional state in this object.

wusto replied:
That's true. The ephemeral key is essential for decrypting the message.

The idea behind this, is a similar interface fore de- and encryption.
The encryption context has the functions getMAC and getEphemeralKey and the decryption context has the function setMAC and setEphemeralKey.

If you like it different, I will change it.

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+
+ECIESCtxBuilder &ECIESCtxBuilder::setKDF(std::shared_ptr<KeyDerivationFunction> kdf)
+{
+    _impl->setKDF(kdf);
```
can be moved

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+ECIESCtxBuilder &ECIESCtxBuilder::setMacFactoryFunction(std::function<std::unique_ptr<MessageAuthenticationCode>
+                                                        (const std::vector<uint8_t> &)> func)
+{
+    _impl->setMacFactoryFunction(func);
```
can be moved

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+ECIESCtxBuilder &ECIESCtxBuilder::setSymmetricCipherFactoryFunction(std::function<std::unique_ptr<SymmetricCipherI>
+                                                                    (const std::vector<uint8_t> &)> func)
+{
+    _impl->setSymmetricCipherFactoryFunction(func);
```
can be moved

wusto replied:
Done.

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+
+std::unique_ptr<ECIESEncryptionCtx> ECIESCtxBuilder::buildEncryptionCtx(AsymmetricPublicKey bobsKey)
+{
+    _impl->_pubKey = std::move(bobsKey);
```
The public code shouldn't access data values of the pimpl class. This has the potential the break ABI stability once again. Please only access it via functions.

wusto replied:
Done

garw commented on src/ecies.cpp:
```diff
@@ -0,0 +1,500 @@
+
+std::unique_ptr<ECIESDecryptionCtx> ECIESCtxBuilder::buildDecryptionCtx(AsymmetricPrivateKey bobsKey)
+{
+    _impl->_privKey = std::move(bobsKey);
```
s.a.

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,400 @@
+ * limitations.
+ * Supported:
+ * - ECSVDP-DH -> Elliptic Curve secret value derivation primitive - Diffie Helman
+ * - ECSVP-DHC (see remark)
```
D missing in ECSVP?

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,400 @@
+ * NOT Supported:
+ * - DLIES (Discrete Logarithm Integrated Encryption Scheme) as this is not ECIES
+ *
+ * If DH or DHC is used depends on the public key.
```
Really? How so?

wusto replied:
Corrected.

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,400 @@
+ *
+ * If DH or DHC is used depends on the public key.
+ *
+ * Remark:
```
This remark doesn't belong into the public documentation of this interface.

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,400 @@
+ *   a cofactor different to h = 1. Therefor DHC is "supported" implicitly. As soon as curves having a cofactor h != 1
+ *   are needed the implementation has to be adapted.
+ *
+ * Elliptic curve integrated encryption scheme provides you the functionality to encrypt a plaintext with an one time
```
```suggestion
 * Elliptic curve integrated encryption scheme provides you the functionality to encrypt a plaintext with a one time
```

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,400 @@
+ *   are needed the implementation has to be adapted.
+ *
+ * Elliptic curve integrated encryption scheme provides you the functionality to encrypt a plaintext with an one time
+ * shared secret (based on an ephemeral key) using a symmetric encryption scheme. All you have to know is:
```
add a remark that the values in brackets are the defaults in this implementation if not changed?

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,400 @@
+ * 2. calculate the shared secret (P_shared = r_eph * P_recipient)
+ * 3. generate the key for symmetric encryption and MAC (key_sym || key_mac = KDF(P_shared)) (using the optional salt)
+ * 4. encrypt the plaintext
+ * 5. calculate the MAC over the ciphertext (with the optional salt)
```
```suggestion
 * 5. calculate the MAC over the ciphertext (with the optional salt concatenated to the ciphertext)
```
?

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,400 @@
+ * 4. encrypt the plaintext
+ * 5. calculate the MAC over the ciphertext (with the optional salt)
+ *
+ * Send P_eph, ciphertext and MAC to the recipient
```
Add a comment that this class does nothing for the serialization of the key or the whole package and it is left to the caller.

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,400 @@
+ * 3. Get the necessary data (private key, MAC scheme, KDF scheme, symmetric cipher scheme, optional salts)
+ * 4. Construct an ECIESDecryptionCtx using the ECIESDecryptionCtxBuilder providing the cipher data (@see
+ *    ECIESEncryptionCtx)
+ * 5. Set the ephemeral key (setEphemeralKey())
```
As this always seems be needed. Why is this separate? Based on our current layout it doesn't look to me like the DecryptionContexts are copyable?

wusto replied:
done.
You are right, they are not copyable because of the internal state.

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,400 @@
+    /**
+     * @brief Set the authentication tag
+     *
+     * As the last step during encryption the authentication tag is calculated. Set this tag to verify the authenticity
+     * of the received message.
```
```suggestion
     * As the last step during encryption the authentication tag is calculated. Set this tag before calling finish()
     * so that the authenticity the received message can be verified.
```

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,400 @@
+};
+
+/**
+ * @brief The ECIESCtxBuilder class is an pseudo abstract class for providing common functions to the Encryption and
```
```suggestion
 * @brief The ECIESCtxBuilder class is a pseudo abstract class for providing common functions to the Encryption and
```

garw replied:
Why is it pseudo abstract? It looks like it could/should be used. This sounds confusing to me.

wusto replied:
The documentation wasn't updated yet.

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,400 @@
+    /**
+     * @brief Sets the required key size for symmetric ciphers
+     *
+     * Default: The default symmetric cipher is AES-CBC PKCS7, thus the key length is 256/8 = 32
```
```suggestion
     * Default: The default symmetric cipher is AES-CBC PKCS7 with a key length of 256/8 = 32
```
Exists as well in 128bit key size. PKCS#7 is the padding format.

wusto replied:
Donne

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,400 @@
+     * This function is a dependency injection.
+     * (e.g. for AES-CBC PKCS7-padding IV=0x00:
+     \code{.cpp}
+        auto cipherDecFunc = [zeroIV](const std::vector<uint8_t> &key) -> std::unique_ptr<AESCipher> {
```
```suggestion
        auto cipherDecFunc = [](const std::vector<uint8_t> &key) -> std::unique_ptr<AESCipher> {
```
same below

wusto replied:
Done

garw commented on src/mococrw/ecies.h:
```diff
@@ -0,0 +1,400 @@
+                    .buildEncryptor();
+        };
+    \endcode
+     * @param func A std::function which returns an object implementing the SymmetricCipherI interface
```
```suggestion
     * @param func A std::function which will be provided with the symmetric key (of key size provided
     *                           in setSymmetricCipherKeySize)  and should return an object  implementing the 
     *                           SymmetricCipherI interface.
```

wusto replied:
Done

garw commented on tests/unit/test_ecies.cpp:
```diff
@@ -0,0 +1,360 @@
+const SymmetricCipherKeySize keySize = SymmetricCipherKeySize::S_256;
+const SymmetricCipherPadding padding = SymmetricCipherPadding::PKCS;
+
+auto macFunc = [](const std::vector<uint8_t> &key) -> std::unique_ptr<MessageAuthenticationCode> {
```
static const?

wusto replied:
Done

garw commented on tests/unit/test_ecies.cpp:
```diff
@@ -0,0 +1,360 @@
+    return std::make_unique<mococrw::HMAC>(openssl::DigestTypes::SHA512, key);
+};
+
+auto randomKey = AsymmetricKeypair(ECCSpec().generate().internal());
```
static const?

wusto replied:
Done

garw commented on src/mococrw/util.h:
```diff
@@ -34,5 +38,43 @@ std::string toHex(const std::vector<uint8_t> &data);
+     *
+     * @param[in] func Function to run on destruction.
+     */
+    Finally(std::function<void()> func)
```
since this is now a general type here:

This should take a universal reference to arbitrary things and std::forward it to the std::function.

wusto replied:
Done

garw commented on src/openssl_wrap.cpp:
```diff
@@ -1325,5 +1325,45 @@ std::vector<uint8_t> _EC_KEY_key2buf(const EVP_PKEY* evp, point_conversion_form_
+std::vector<uint8_t> _EVP_derive_key(const EVP_PKEY *peerkey, const EVP_PKEY *key)
+{
+    size_t secret_len;
+    /* EVP_PKEY_CTX_new: Why const_cast is allowed:
```
The real question should be: Why is const_cast needed here and ok? Why do these contexts want mutable objects in the first place?

wusto replied:
Done

garw commented on src/openssl_wrap.cpp:
```diff
@@ -1325,5 +1325,45 @@ std::vector<uint8_t> _EC_KEY_key2buf(const EVP_PKEY* evp, point_conversion_form_
+    /* Derive the shared secret */
+    OpensslCallIsOne::callChecked(lib::OpenSSLLib::SSL_EVP_PKEY_derive, ctx.get(), result.data(), &secret_len);
+
+    /* Never use a derived secret directly. Typically it is passed
```
This information is here rather useless. If you want people to read this, put it into the function description.

wusto replied:
Done

garw commented on tests/unit/test_asymmetric_encryption.cpp:
```diff
@@ -448,3 +448,12 @@ TEST_F(AsymmetricEncryptionTest, testUnsuccessfulDecryptionWrongLabelOEAP)
     EXPECT_THROW(ctxDecrypt.decrypt(OAEPPaddingDataSet.encrypted), MoCOCrWException);
 }
+
+TEST(AsymmetricPublicKeyTest, createEphemeralKey)
```
```suggestion
TEST(AsymmetricPublicKeyTest, ecdhSmokeTest)
```

wusto replied:
done

garw commented on tests/unit/test_asymmetric_encryption.cpp:
```diff
@@ -448,3 +448,12 @@ TEST_F(AsymmetricEncryptionTest, testUnsuccessfulDecryptionWrongLabelOEAP)
+    AsymmetricKey key(spec.generate());
+    AsymmetricPublicKey pubKey = spec.generate().internal();
+    /* Just check if the function returns without indicating a failure */
+    openssl::_EVP_derive_key(pubKey.internal(), key.internal().get());
```
EXPECT_NO_THROW?

wusto replied:
Done

garw commented on src/key.cpp:
```diff
@@ -91,6 +102,35 @@ AsymmetricPublicKey AsymmetricPublicKey::readPublicKeyFromPEM(const std::string&
+AsymmetricPublicKey AsymmetricPublicKey::fromECPoint(const std::shared_ptr<ECCSpec> keySpec,
+                                                     const std::vector<uint8_t> &point)
+{
+    ensureEccCurveSupported(keySpec.get()->curve());
```
```suggestion
    ensureEccCurveSupported(keySpec->curve());
```
?!

wusto replied:
Done

garw commented on src/key.cpp:
```diff
@@ -91,6 +102,35 @@ AsymmetricPublicKey AsymmetricPublicKey::readPublicKeyFromPEM(const std::string&
+
+    SSL_EC_KEY_Ptr ec_key = _EC_KEY_oct2key(static_cast<int>(keySpec.get()->curve()), point);
+    SSL_EVP_PKEY_Ptr evp_key(_EVP_PKEY_new());
+    /* As set1 is incrementing the ref count, we can invoke the destructor of ec_key here */
```
```suggestion
    /* As set1 is incrementing the internal ref count, we can and need to invoke the destructor of ec_key here */
```

wusto replied:
Done

garw reviewed (changes requested):
Please have a look again and also fix the tiny little details, in particular if the public interface is involved. It is almost impossible to change this in the future.

garw reviewed (approved):
looks good now. Can be merged.
